### DISCUSSION TITLE
What is the correct way to set a username in the registration flow?

### DISCUSSION THREAD

**Discussion by christian-roggia** [ory/kratos#983] [Q&A] [answered]
Hello,

I would like to get a bit of clarification regarding usernames in the registration flow:

- What is the correct flow to sanitize the username and make sure it satisfies a specific regex (e.g. only alphanumeric characters)?
- How do I sanitize usernames coming from third-party OIDC like Google or GitHub?
- How can I verify beforehand that the username is available during registration?
- How do I append a random suffix (e.g. a number) to OIDC usernames if the username returned by a provider is already taken?

Thank you!

**Comment by aeneasr** [comment, marked as answer]
> What is the correct flow to sanitize the username and make sure it satisfies a specific regex (e.g. only alphanumeric characters)?

In the identity json schema using https://json-schema.org/understanding-json-schema/reference/string.html#regular-expressions

> How do I sanitize usernames coming from third-party OIDC like Google or GitHub?

You write jsonnet code: https://jsonnet.org

> How can I verify beforehand that the username is available during registration?

You can't right now. It would allow account enumeration attacks.

> How do I append a random suffix (e.g. a number) to OIDC usernames if the username returned by a provider is already taken?

Not sure if jsonnet has a rand function built into the stdlib, but it does have the md5 function. You can probably use that

**Comment by christian-roggia** [comment]
I would like to share the results of our research and tests:

- OIDC and support for multiple email addresses work together
- The best way to set usernames is to let the users choose them during the registration flow rather than pre-filling them (our opinion about it)
- Make sure that your regex in the JSON schema is as simple as possible, we couldn't introduce regex lookahead even if they are supposed to be supported by version 0.7 as per official documentation
- Twitter OIDC sucks and doesn't support OAuth 2.0

This is an example of .jsonnet for multiple email addresses with Google:
```jsonnet
local claims = {
  email_verified: false
} + std.extVar('claims');

{
  identity: {
    traits: {
      # Verified email address.
      [if "email" in claims && claims.email_verified then "emails" else null]: [ claims.email ],
    },
  },
}
```

This is a fragment of how our JSON schema looks like:
```json
...
        "username": {
          "type": "string",
          "pattern": "^[a-z0-9]\\.?(([a-z0-9]+)\\.?)+[a-z0-9]$",
          "minLength": 4,
          "maxLength": 20,
          "title": "Username",
          "ory.sh/kratos": {
            "credentials": {
              "password": {
                "identifier": true
              }
            }
          }
        }
...
```

NOTE: We support only usernames that are between 4 and 20 characters, that are composed of lowercase letters, digits, and dots, with no repeating dots and no leading or trailing dots.

---
@aeneasr before marking the question as answered, I wanted to ask a clarification about my last question:
**is it possible to know inside the jsonnet logic whether the username has already been taken or not and act accordingly?**

Thank you, btw there is no random function in jsonnet as it is designed to be hermetic FYI.

**Reply by aeneasr** [reply]
> is it possible to know inside the jsonnet logic whether the username has already been taken or not and act accordingly?

Not right now unfortunately. We could probably add such a look up function though. I assume you are really talking about usernames here, not about emails?

> Thank you, btw there is no random function in jsonnet as it is designed to be hermetic FYI.

Ahh, unfortunate - we can definitely add a random function though to the set of non-stdlib functions! PRs welcomed :)

**Comment by christian-roggia** [comment]
> Not right now unfortunately. We could probably add such a look up function though. I assume you are really talking about usernames here, not about emails?

Yes, I am talking about usernames and it would be really helpful to have that information inside the mapper. I would guess it would be in general useful to be able to fetch any arbitrary information from the identity schema via jsonnet.

> Ahh, unfortunate - we can definitely add a random function though to the set of non-stdlib functions! PRs welcomed :)

That would be really helpful for some use-cases like ours.

I am gonna open a few tickets and mark this question as answered.

**Reply by aeneasr** [reply]
> Yes, I am talking about usernames and it would be really helpful to have that information inside the mapper. I would guess it would be in general useful to be able to fetch any arbitrary information from the identity schema via jsonnet.

Yeah that makes sense to me. We could probably start collecting some functions that people have access to in the JsonNet code. Open for suggestions :) So far we have:

- rand
- verifiableEmailAddressKnown
- recoveryEmailAddressKnown
- loginMethodPasswordIdentifierKnown

**Reply by christian-roggia** [reply]
Wouldn't it be more flexible to allow the JSON flow to fetch any arbitrary data directly without having hardcoded functions?
I am thinking about something like:

```jsonnet
local is_username_taken = ory.isValueKnown("traits.username", "myusername")
local is_email_known = ory.isValueKnown("traits.email", "abc@example.com")
local is_email_array_known = ory.isValueArrayKnown("traits.emails", "abc@example.com") # if multiple emails are supported
local is_my_arbitrary_field_known = ory.isValueKnown("traits.myfield", "myvalue")
local is_my_arbitrary_field_array_known = ory.isValueArrayKnown("traits.myarray", "myarrayvalue")
```

**Reply by aeneasr** [reply]
Unfortunately not because we don't index the traits fields. It would be quite expensive to do so. Loading and looping over them is also way too expensive and slow for large datasets. The hardcoded functions above are all using indexed fields!

**Comment by EiffelFly** [comment]
Hi @aeneasr 

In our user case we need this feature too. (We have a unique username that currently stored in our database but we want to have a way store this kind of information directly in the kratos and have a way to verified it)

Is there has any solution around this issue?
